crates/sqlexec/src/remote/location_aware.rs

```diff
@@ -0,0 +1,109 @@
+use crate::planner::physical_plan::client_recv::ClientExchangeRecvExec;
+use async_trait::async_trait;
+use datafusion::error::Result;
+use datafusion::{
+    arrow::datatypes::SchemaRef,
+    datasource::TableProvider,
+    execution::context::SessionState,
+    logical_expr::{LogicalPlan, TableProviderFilterPushDown, TableType},
+    physical_plan::{ExecutionPlan, Statistics},
+    prelude::Expr,
+};
+use std::{any::Any, sync::Arc};
+use uuid::Uuid;
+
+/// Who should be scanning the table.
+#[derive(Debug, Clone, Copy)]
+pub enum ScanLocation {
+    /// Client should scan the table (e.g. local CSV files or data frames).
+    ///
+    /// This will require a data exchange.
+    Client {
+        /// ID to use when constructing the recv exec.
+        ///
+        /// When the server begins executing, it will attempt to pull from the
+        /// recv exec. The client will be responsible for sending record batches
+        /// to the server using this ID.
+        broadcast_id: Uuid,
+    },
+    /// Server should scan the table.
+    Server,
+}
```

Comment:
I was thinking doing it in our dispatch code since the rules are pretty simple. If table is a local file scan or data frame, set scan location to client. Otherwise set to server.

We'll need additional logic in this `scan` function to avoid creating the client exec if we're not actually connected to a remote node.

Reply:
that makes sense. as of right now, i cant think of any explicit reasons that logic would **need** to be in the planner.  Since that's probably the simplest, i'd say go with it. 

However...

 I do think longterm it would be the most fitting in the physical/extension planner. I don't think a logical plan should know anything about client/server relationships.

Reply:
I generally agree with wanting this in the planner, and I looked a bit at this and see two road blocks:

1. If we continue returning `Arc<dyn TableProvider>`, we'd need to try to downcast to every possible table that should be scanned by the client. And while we only care about data frames and local file scans for that, we'd need to make sure that both have a unique and well-known type. The data frame would be easy, but local scanning would be a bit trickier, since it'd use the `ObjStoreTableProvider` type, which is _also_ used for s3 and gcs.
2. The default planner doesn't give us a good extension point for custom table sources, so to swap out the providers, we'd need to walk the plan prior to physical planning

And now that I'm writing this out, I think both of these are pretty small roadblocks that are easily solvable... I'll try to avoid littering the code with `LocationAwareTableProvider` and see if that can all be centralized to planning.